File: 2018/06.py
from lib import disp, rl
import math, copy, re, hashlib
import itertools as it


def man_dist(l1, l2):
	return abs(l1[0] - l2[0]) + abs(l1[1] - l2[1])

def get_bordered(g):
	border = set()
	l = len(g) - 1
	for i in rl(g):
		border |= {g[0][i], g[l][i], g[i][0], g[i][l]}
	return list(border)

def get_max_area(g, border):
	areas = dict()
	for i in rl(g):
		for j in rl(g[i]):
			l = g[i][j]
			if l == '.' or l in border: continue
			try:
				areas[l] += 1
			except:
				areas[l] = 1
	maxi = 0
	for k in areas.keys():
		if areas[k] > maxi:
			maxi = areas[k]
	return maxi

def part_1(data):
	GRID = [["." for i in range(400)] for j in range(400)]

	lets = [chr(i) for i in range(ord('a'), ord('z') + 1)] + [chr(i) for i in range(ord('A'), ord('Z') + 1)]
	li = 0
	# array of tuples (letter, x-coord, y-coord)
	locations = []
	for coords in data:
		x, y = map(int, coords.split(', '))
		locations.append((lets[li], x, y))
		GRID [y][x] = lets[li]
		li += 1
	lets = lets[:li]
	li = 0

	for i in rl(GRID):
		for j in rl(GRID[i]):
			mini = 400 ** 2
			s_mini = None
			closest = None
			for loc in locations:
				l, x, y = loc
				if (d:=man_dist((i, j), (y, x))) <= mini:
					s_mini = mini
					mini = d
					closest = l
			if s_mini != mini:
				GRID[i][j] = closest

	border = get_bordered(GRID)
	maxi = get_max_area(GRID, border)

	print(maxi)
	print('END OF PART1')
	return

def is_under_n(coord, n, locs):
	suma = 0
	for l in locs:
		suma += man_dist(coord, l)

	return suma < n


def part_2(data):
	GRID = [["." for i in range(800)] for j in range(800)]
	locations = []
	for coords in data:
		x, y = map(lambda x: int(x) + 200, coords.split(', '))
		locations.append((x, y))

	cnt = 0
	n = 10000
	for i in rl(GRID):
		for j in rl(GRID[i]):
			if is_under_n((i, j), n, locations):
				cnt += 1

	print(cnt)

	print('END OF PART2')
	return 


if __name__ == '__main__':
	with open('06_input') as f:
		data = f.read()
		data = data.split('\n')
		# data = list(map(int, data.split()))


	# part_1(copy.deepcopy(data))
	part_2(copy.deepcopy(data))
	

File: 2018/06.js

var fs = require('fs');
const { ord, chr, debug, disp, int, float, randint, gcd, lcm, crt, modPow, mod } = require("../lib");
const { range, drange, trange, iter, diter, titer } = require("../lib");
const { PriorityQueue, Queue, set, Stack } = require("../lib");
const { min, max, random, abs, ceil, floor, log, log10, log2, round, sign, sin, cos, tan, asin, acos, atan, atan2, sqrt, PI } = Math;
const { isSuperset, or, and, xor, sub } = set;

const sort = (arr, fn = (a, b) => a-b) => {
	arr.sort(fn);
}
const in_ = (a, arr) => arr.indexOf(a) !== -1;

const D4 = [[0,1],[1,0],[0,-1],[-1,0]];
const D8 = [[0,1],[1,1],[1,0],[1,-1],[0,-1],[-1,-1],[-1,0],[-1,1]];
const MOD = 1e9+7;

const parse = (data) => {
	return data.map(r => r.split(',').map(a => int(a.trim())));
}

const manDist = (c1, c2) => {
	const [x1, y1] = c1, [x2, y2] = c2;
	return abs(x1 - x2) + abs(y1 - y2);
}

function part1(data) {
	data = parse(data);

	const G = new Array(400).fill(0).map(n => new Array(400).fill(0));

	drange(400)((i, j) => {
		let mn = 1e9;
		let mni = -1;
		data.map((coords, k) => {
			const d = manDist(coords, [i,j]);
			if (d <= mn) {
				mn = d;
				mni = k;
			}
		});
		G[i][j] = mni;
	});

	const used = new set(new Array(data.length).fill(0).map((_, i) => i));
	range(400)(i => {
		used.delete(G[0][i]);
		used.delete(G[400 - 1][i]);
		used.delete(G[i][0]);
		used.delete(G[i][400 - 1]);
	});

	const area = new Array(data.length).fill(0);
	// disp(G)
	drange(400)((i, j) => {
		if (used.has(G[i][j])) {
			area[ G[i][j] ]++;
		}
	});

	debug(max(...area))

	console.log("END OF PART1");
	return;
}

const dfs = (data, MX, start) => {
	const q = new Queue();
	const s = new set();
	q.push(start);

	while (!q.empty()) {
		const [x, y] = q.pop();
		if (s.has([x, y])) {
			continue;
		}
		let sm = 0;
		iter(data)(d => {
			sm += manDist(d, [x, y]);
		});
		if (sm >= MX) {
			continue;
		}
		s.add([x, y]);
		for (d of D4) {
			const [dx, dy] = d;
			q.push([x+dx,y+dy]);
		}
	}
	return s;
}

function part2(data) {
	data = parse(data);
	const MX = 10_000;

	const start = data.reduce((acc, curr) => {
		acc[0] += curr[0];
		acc[1] += curr[1];
		return acc;
	}, [0, 0]);
	start[0] = Math.floor(start[0] / data.length);
	start[1] = Math.floor(start[1] / data.length);

	const G = dfs(data, MX, start);
	debug(G.size)
	console.log("END OF PART2");
	return;
}

function main() {
	let data = fs.readFileSync("06_input").toString("utf-8");
	data = data.split('\n');
	// data = data.split('\n').map(a => Number(a));

	if (Array.isArray(data)) {
		part1(Array.from(data));
		part2(Array.from(data));
	} else {
		part1(data);
		part2(data);
	}
}

main();
